### ISSUE TITLE
ParallelTestExecution with Sorted output and Tags sometimes skips tests

### ISSUE THREAD

**Issue by maiflai** [scalatest/scalatest#706]
Hi,

This follows from https://github.com/maiflai/gradle-scalatest/issues/25.

```scala
import org.scalatest.tagobjects.{CPU, Network}
import org.scalatest.{FlatSpec, ParallelTestExecution}

class MySpec extends FlatSpec with ParallelTestExecution {
  it should "network" taggedAs Network in {
    fail("doesn't")
  }
  it should "cpu" taggedAs CPU in {
    fail("doesn't")
  }
}
```
```bash
java org.scalatest.tools.Runner -oF -PS4 -R build/classes/test -n org.scalatest.tags.CPU
```
```bash
Discovery starting.
Discovery completed in 94 milliseconds.
Run starting. Expected test count is: 1
Run completed in 202 milliseconds.
Total number of tests run: 0
Suites: completed 1, aborted 0
Tests: succeeded 0, failed 0, canceled 0, ignored 0, pending 0
No tests were executed.
```
It correctly discovers that one test should be run, but then appears not to run it. As you can see from the test code, running the test should fail.

Changing `-PS4` to `-P4` always runs the expected test and fails the build.

Removing `-n org.scalatest.tags.CPU` detects and runs two tests. Both fail.

Replacing `-n org.scalatest.tags.CPU` with `-l org.scalatest.tags.CPU` should run the other test, but actually gives the same `No tests were executed`

I've tried picking through `Runner.scala` but got somewhat lost in the middle. Is this a recognised issue that has a known workaround please?

Thanks,
Stu

**Comment by jeksmith** [comment]
We're hitting this problem. When we use -l to exclude our slow tests, it skips running tests that aren't marked slow when -PS is used. If just -P is used, all tests run. This is a problem in 3.1.1, 3.1.2, and 3.2.0. For us, we have 32 "fast" tests in one of our projects and get this output:

`Discovery starting.
Discovery completed in 3 seconds, 483 milliseconds.
Run starting. Expected test count is: 32
Run completed in 4 seconds, 518 milliseconds.
Total number of tests run: 17
Suites: completed 9, aborted 0
Tests: succeeded 17, failed 0, canceled 0, ignored 0, pending 0`

Only 17 run, though 32 were actually expected.

Simply changing from -PS12 to -P12 we get:

`Discovery starting.
Discovery completed in 3 seconds, 524 milliseconds.
Run starting. Expected test count is: 32
Run completed in 4 seconds, 507 milliseconds.
Total number of tests run: 32
Suites: completed 16, aborted 0
Tests: succeeded 32, failed 0, canceled 0, ignored 1, pending 0
`

We're stuck with -PS, so any help would be appreciated greatly. Thanks!

**Comment by maiflai** [comment]
Adding `-Dorg.scalatest.tools.Runner.forever=true` seems to trigger the example test above to actually run and fail (infinitely).

I'll keep digging.

**Comment by jeksmith** [comment]
Thanks for looking @maiflai. FWIW, I added a 5-second sleep to one of our tests that isn't running on my machine and then all of the tests got executed. However, when I instead added a 5-second sleep to one of the tests that were already running (one of the 17), then no change in the number of tests executed (and it took 5 seconds longer). It doesn't seem to be dependent on the test itself, as a co-worker has a different set of the 32 tests that get executed than mine (and it is only 10).

**Comment by cheeseng** [comment]
@maiflai @jeksmith @bvenners I submitted the following PR attempting to fix this: 

https://github.com/scalatest/scalatest/pull/1838

**Comment by maiflai** [comment]
that's great, thanks very much.

the patch appearedfixed the issue when I applied it locally.

**Comment by jeksmith** [comment]
@cheeseng Unfortunately the patch didn't work for my case. Perhaps it is different when using -l alone instead of -n alone? I verified I am using your change. I'm not sure how to boil down my suite of tests to a small example that exposes the problem. Any ideas of what I should look for in naming or otherwise?

Thanks for your help with this!

**Comment by cheeseng** [comment]
@jeksmith I afraid I am not sure how I can help you in your case, a small example project to reproduce the problem will be really helpful, the above fix was identified with @maiflai 's example.  It could be a separate problem for your case, and it will be really hard to tackle your problem without a consistent way to reproduce it, so hopefully you can somehow get a boil down version that reproduce your problem and I'll be happy to look into it.  :)

**Comment by jeksmith** [comment]
@cheeseng I've been working on narrowing it down. I've been deleting tests in an effort to find the part that causes trouble. The patch does run a few more tests in most cases, but still misses a lot of them. I hope to have something to you in a day or so.

**Comment by jeksmith** [comment]
@cheeseng Ok, I found the test class that causes trouble for the PR. It is an AsyncFunSpec. I've boiled it down to the following. When running with @maiflai 's class and excluding Slow, it doesn't run any test about 50% of the time.

```
import org.scalatest.funspec.AsyncFunSpec
import org.scalatest.tagobjects.Slow
import scala.concurrent.Future

class MyAsyncSpec extends AsyncFunSpec {
  describe("when run in parallel with other tests") {
    it("should not affect their execution", Slow) {
      Future { assert(false) }
    }
  }
}
```

Use `-l  org.scalatest.tags.Slow` (and not `-n`). 

For my actual test, it doesn't work 100% of the time. For this test, like I said, it doesn't work about 50% of the time. By doesn't work, I mean when you run it with the MySpec class above, the CPU and Network tests are not executed as they should be since they are NOT Slow tests.

**Comment by maiflai** [comment]
I think that `AsyncTestSuite` has special behaviour for two mixins (`org.scalatest.AsyncTestSuite#parallelAsyncTestExecution`)?

`class MyAsyncSpec extends AsyncFunSpec with ParallelTestExecution` seems to work reliably.

`class MyAsyncSpec extends AsyncFunSpec with RandomTestOrder` does not.

**Comment by cheeseng** [comment]
@maiflai @jeksmith Fyi @johnduffell has a more complete fix here for 3.2 and 3.0: 

https://github.com/scalatest/scalatest/pull/1842
https://github.com/scalatest/scalatest/pull/1841

I have pushed similar changes to my current PR for 3.1.x: 

https://github.com/scalatest/scalatest/pull/1838

Do you mind to test again and see if the changes fix your problem now?

Thanks.

**Comment by maiflai** [comment]
Yes, that seems to work, and I agree those are the uses of that constructor in 3.2

Thanks very much!

Stu

**Comment by jeksmith** [comment]
@cheeseng Excellent! Works in my test environment and with my dev project. All tests are executed as expected for both '-n' and '-l' cases. Many thanks!!!!

    -John

**Comment by jeksmith** [comment]
@cheeseng Hi. Any chance you know when the fix may get into a Scalatest release? Thanks.
